php-checks/src/main/java/org/sonar/php/checks/LoopExecutingAtMostOnceCheck.java

```diff
@@ -0,0 +1,151 @@
+import org.sonar.plugins.php.api.tree.statement.BreakStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ForEachStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ForStatementTree;
+import org.sonar.plugins.php.api.tree.statement.GotoStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ReturnStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ThrowStatementTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+import org.sonar.plugins.php.api.visitors.PreciseIssue;
+
+import static org.sonar.php.checks.utils.SyntacticEquivalence.areSyntacticallyEquivalent;
+import static org.sonar.php.tree.TreeUtils.findAncestorWithKind;
+import static org.sonar.php.tree.TreeUtils.isDescendant;
+
+@Rule(key = LoopExecutingAtMostOnceCheck.KEY)
+public class LoopExecutingAtMostOnceCheck extends PHPVisitorCheck {
+
+  public static final String KEY = "S1751";
+  private static final String MESSAGE = "Refactor this loop to do more than one iteration.";
+  private static final Set<Tree.Kind> LOOPS = Sets.immutableEnumSet(
+    Tree.Kind.WHILE_STATEMENT,
+    Tree.Kind.DO_WHILE_STATEMENT,
+    Tree.Kind.FOR_STATEMENT,
+    Tree.Kind.FOREACH_STATEMENT);
+
+  private ListMultimap<Tree, Tree> reportedLoops = ArrayListMultimap.create();
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    reportedLoops.clear();
+    super.visitCompilationUnit(tree);
+    reportIssues();
+  }
+
+  private void reportIssues() {
+    reportedLoops.asMap().forEach((loop, jumps) -> {
+      PreciseIssue preciseIssue = context().newIssue(this, ((PHPTree) loop).getFirstToken(), MESSAGE);
+      jumps.forEach(jump -> preciseIssue.secondary(jump, "loop exit"));
+    });
+  }
+
+  @Override
+  public void visitBreakStatement(BreakStatementTree tree) {
+    checkJump(tree);
+  }
+
+  @Override
+  public void visitReturnStatement(ReturnStatementTree tree) {
+    checkJump(tree);
+  }
+
+  @Override
+  public void visitThrowStatement(ThrowStatementTree tree) {
+    checkJump(tree);
+  }
+
+  @Override
+  public void visitGotoStatement(GotoStatementTree tree) {
+    checkJump(tree);
+  }
+
+  private void checkJump(Tree tree) {
+    Tree loop = findAncestorWithKind(tree, LOOPS);
+    if (loop != null && !canExecuteMoreThanOnce(loop) && !isForEachReturningFirstElement(loop, tree)) {
+      reportedLoops.put(loop, tree);
+    }
+  }
+
+  private static boolean isForEachReturningFirstElement(Tree loop, Tree jump) {
+    if (!loop.is(Tree.Kind.FOREACH_STATEMENT) || !jump.is(Tree.Kind.RETURN_STATEMENT)) {
+      return false;
+    }
+    ReturnStatementTree returnTree = (ReturnStatementTree) jump;
+    ForEachStatementTree forEachTree = (ForEachStatementTree) loop;
+    return areSyntacticallyEquivalent(returnTree.expression(), forEachTree.value())
+      || areSyntacticallyEquivalent(returnTree.expression(), forEachTree.key());
+  }
+
+  private boolean canExecuteMoreThanOnce(Tree loop) {
```

Comment:
can we cache result of this function? if loop has 10 returns, we will build CFG 10 times, while it's easy to maintain a map per file `Map<Loop Tree, Boolean>`

Reply:
I am not sure about this. I hesitate to add such ad-hoc caching in specific rule. If we want to cache, let's cache whole CFG for given function. I think having more loops in the same function is more common scenario than having multiple jumps in the single loop.